server/chat.ts

```diff
@@ -1293,509 +1088,738 @@ export class CommandContext extends MessageContext {
+			// \u00ae\u00a9 (R) (C)
+			// tslint:disable-next-line: max-line-length
+			if (nameSymbols.length > 4 || /[^a-z0-9][a-z0-9][^a-z0-9]/.test(name.toLowerCase() + ' ') || /[\u00ae\u00a9].*[a-zA-Z0-9]/.test(name)) name = name.replace(/[\u00A1-\u00BF\u00D7\u00F7\u02B9-\u0362\u2012-\u2027\u2030-\u205E\u2050-\u205F\u2190-\u23FA\u2500-\u2BD1\u2E80-\u32FF\u3400-\u9FFF\uF900-\uFAFF\uFE00-\uFE6F]+/g, '').replace(/[^A-Za-z0-9]{2,}/g, ' ').trim();
+		}
+		name = name.replace(/^[^A-Za-z0-9]+/, ""); // remove symbols from start
+		name = name.replace(/@/g, ""); // Remove @ as this is used to indicate status messages
+
+		// cut name length down to 18 chars
+		if (/[A-Za-z0-9]/.test(name.slice(18))) {
+			name = name.replace(/[^A-Za-z0-9]+/g, "");
+		} else {
+			name = name.slice(0, 18);
 		}
-		toUncache = newuncache;
-	} while (toUncache.length > 0);
-}
 
-function uncacheDir(root: string) {
-	const absoluteRoot = FS(root).path;
-	for (const key in require.cache) {
-		if (key.startsWith(absoluteRoot)) {
-			delete require.cache[key];
+		name = Dex.getName(name);
+		for (const curFilter of Chat.namefilters) {
+			name = curFilter(name, user);
+			if (!name) return '';
 		}
+		return name;
 	}
-}
 
-function uncache(path: string) {
-	const absolutePath = require.resolve('../' + path);
-	delete require.cache[absolutePath];
-}
+	hostfilters: HostFilter[] = [];
+	hostfilter(host: string, user: User, connection: Connection, hostType: string) {
+		for (const curFilter of Chat.hostfilters) {
+			curFilter(host, user, connection, hostType);
+		}
+	}
 
-function loadPlugins() {
-	if (Chat.commands) return;
+	loginfilters: LoginFilter[] = [];
+	loginfilter(user: User, oldUser: User | null, usertype: string) {
+		for (const curFilter of Chat.loginfilters) {
+			curFilter(user, oldUser, usertype);
+		}
+	}
 
-	// tslint:disable-next-line: no-floating-promises
-	FS('package.json').readIfExists().then(data => {
-		if (data) Chat.packageData = JSON.parse(data);
-	});
+	nicknamefilters: NameFilter[] = [];
+	nicknamefilter(nickname: string, user: User) {
+		for (const curFilter of Chat.nicknamefilters) {
+			nickname = curFilter(nickname, user);
+			if (!nickname) return '';
+		}
+		return nickname;
+	}
 
-	// prevent TypeScript from resolving
-	const commandsFile = '../server/chat-commands';
-	Chat.baseCommands = require(commandsFile).commands;
-	Chat.basePages = require(commandsFile).pages;
-	Chat.commands = Object.assign({}, Chat.baseCommands);
-	Chat.pages = Object.assign({}, Chat.basePages);
+	statusfilters: StatusFilter[] = [];
+	statusfilter(status: string, user: User) {
+		status = status.replace(/\|/g, '');
+		for (const curFilter of Chat.statusfilters) {
+			status = curFilter(status, user);
+			if (!status) return '';
+		}
+		return status;
+	}
+	/*********************************************************
+	 * Translations
+	 *********************************************************/
+	// language id -> language name
```

Comment:
```suggestion
	/** language id -> language name */
```

Documenting properties should be in doc comments so they can show up on hover.